lib/connect.js

```diff
@@ -234,25 +240,86 @@ const getTasks = (model, setupListeners) => {
         setupListeners(mongoClient);
       }
 
+      let maybeTunnelError;
+
+      if (tunnel) {
+        tunnel.on('error', (tunnelError) => {
+          if (tunnelError.origin === 'ssh-client') {
+            maybeTunnelError = tunnelError;
+          }
+        });
+      }
+
       mongoClient.connect((err, _client) => {
+        // When mongo client failed to connect AND tunnel encountered an error
+        // those two are totally related and we want to return a tunnel error as
+        // it will have more relevant context for the end user
+        err = err && maybeTunnelError ? maybeTunnelError : err;
+
         ctx(err);
 
         if (err) {
           if (tunnel) {
             debug('data-service connection error, shutting down ssh tunnel');
-            tunnel.close();
+            tunnel.close().then(
+              () => {
+                debug('ssh tunnel stopped');
+                cb(err);
+              },
+              (tunnelError) => {
+                debug('ssh tunnel stopped with error: %s', tunnelError.message);
+                cb(err);
+              }
+            );
           }
-
-          return cb(err);
+          return;
         }
 
         client = _client;
 
         if (tunnel) {
-          client.on('close', () => {
-            debug('data-service disconnected. shutting down ssh tunnel');
-            tunnel.close();
-          });
+          const origClose = client.close;
```

Comment:
Could we do the same without tampering with the mongo client, what about MongoClient events?

Reply:
This is what I mentioned in PR description when I said that other options will not give us enough control there or will leak too many implementation details outside of `connection-model`:

We can't use `MongoClient` events, this would basically be the current implementation that just fires `tunnel.close` on client `close` event. This does not guarantee that tunnel is closed before `client.close` callback is fired and this is what causes the race issue when re-connecting quickly currently.

The only other option I see is to expose tunnel as part of the connect method callback and then make whoever is using the `connect` method (`data-service`) to take care of cleaning up of everything on disconnect manually, but IMO it exposes too much implementation details about the ssh connection options to `connection-model` consumers, this is why I didn't went this way

Reply:
I think modifying the client is too unclear. What about returning a `disconnect` method?

return [connectedClient, disconnect] ? 

like that the dataservice wouldn't need to know about the tunnel but we also don't mess with the MongoClient

Reply:
Discussed that we will return client and tunnel and done in 7e1ac00